File: examples/socketcan/nodejs/can_send.js
#!/usr/bin/env node
// Transmit a CAN frame on a rexgend SocketCAN (vcan) interface.
//   usage: node can_send.js [bus] [ID_hex] [hexdata]
//     e.g. node can_send.js can0 153 AABBCCDDEEFF0011
//          node can_send.js can0 1ABCDEF0 DEADBEEF        # extended ID
//
// Requires the "socketcan" package:  npm install socketcan
// Needs use_socketcan=1 in rexgend.conf.

let can;
try {
  can = require('socketcan');
} catch (e) {
  console.error('missing dependency: run "npm install socketcan"');
  process.exit(1);
}

const iface = process.argv[2] || 'can0';
const idStr = process.argv[3] || '153';
const hex = process.argv[4] || 'AABBCCDDEEFF0011';

const id = parseInt(idStr, 16);
const ext = id > 0x7ff || idStr.length > 3;     // extended if >11 bits / >3 hex digits
const data = Buffer.from(hex, 'hex');

let channel;
try {
  channel = can.createRawChannel(iface);
  channel.start();
  channel.send({ id, ext, rtr: false, data });
} catch (e) {
  console.error(`cannot send on ${iface} (use_socketcan=1 and ${iface} configured?) - ${e.message}`);
  process.exit(1);
}

console.log(`sent on ${iface}: id=0x${id.toString(16).toUpperCase()} ${ext ? '(ext) ' : ''}data=${hex.toUpperCase()}`);
channel.stop();

File: examples/socketcan/python/can_send.py
#!/usr/bin/env python3
"""Transmit a CAN frame on a rexgend SocketCAN (vcan) interface.

usage: ./can_send.py [bus] [ID_hex] [byte ...]
   e.g. ./can_send.py can0 153 AA BB CC DD EE FF 00 11
        ./can_send.py can0 1ABCDEF0 DE AD BE EF          # extended ID

Standard library only. Needs use_socketcan=1 in rexgend.conf.
"""
import socket
import struct
import sys

bus = sys.argv[1] if len(sys.argv) > 1 else "can0"
can_id = int(sys.argv[2], 16) if len(sys.argv) > 2 else 0x153
data = bytes(int(b, 16) for b in sys.argv[3:]) if len(sys.argv) > 3 \
    else bytes([0xAA, 0xBB, 0xCC, 0xDD, 0xEE, 0xFF, 0x00, 0x11])

CAN_FRAME_FMT = "=IB3x8s"          # struct can_frame, 16 bytes
CAN_EFF_FLAG = 0x80000000

# Extended ID if it does not fit in 11 bits, or was given as >3 hex digits.
if can_id > 0x7FF or (len(sys.argv) > 2 and len(sys.argv[2]) > 3):
    can_id |= CAN_EFF_FLAG

dlc = min(len(data), 8)
frame = struct.pack(CAN_FRAME_FMT, can_id, dlc, data[:8].ljust(8, b"\x00"))

try:
    s = socket.socket(socket.PF_CAN, socket.SOCK_RAW, socket.CAN_RAW)
    s.bind((bus,))
    s.send(frame)
except OSError as e:
    sys.exit(f"cannot send on {bus} (use_socketcan=1 and {bus} configured?) - {e}")

print(f"sent on {bus}: id=0x{can_id & 0x1FFFFFFF:X} dlc={dlc} "
      f"data={data[:dlc].hex(' ').upper()}")
